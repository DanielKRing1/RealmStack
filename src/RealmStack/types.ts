import { Dict } from '@asianpersonn/dict-utils/dist/types';
import Realm from 'realm';

import { COLUMN_NAME_SNAPSHOT_TIMESTAMP } from "../constants/naming";

export type RSCreateParams = {
    snapshotProperties: Dict<any>;
} & RSLoadableParams;

export type RSLoadStackParams = {
    shouldReloadRealm?: boolean;
} & RSLoadableParams;

export type RSPushParams = {
    timestamp?: Date;
};

export type RSLoadableParams = {
    metaRealmPath: string;
    loadableRealmPath: string;
    stackName: string;
}

export type RSDeleteStackParams = {
    reloadRealm: () => Promise<Realm>;
} & RSLoadableParams;

export type RSUpdatePropertiesParams = {
    newSnapshotProperties: Dict<any>;

    reloadRealm: () => Promise<Realm>;
} & RSLoadableParams;

export type RealmStack = {
    getStackName: () => string;
    getMetaRealmPath: () => string;
    getLoadableRealmPath: () => string;
    getPropertyNames: () => string[];
    getProperties: () => Dict<any>;

    getAllSnapshots: () => Promise<Realm.Results<RSSnapshot & Realm.Object>>;
    getStackRow: () => Promise<(RealmStackRow & Realm.Object) | undefined>;
    getListJSON: () => Promise<RSSnapshot[]>;

    deleteStack: () => Promise<void>;
    updateStack: (newSnapshotProperties: Dict<any>) => Promise<void>;

    push: (args: RSPushParams, ...snapshots: Dict<any>[]) => Promise<boolean>;
    getClosestDate: (searchDate: Date) => Promise<number>;

    loadRealm: () => Promise<Realm>;
    reloadRealm: () => Promise<Realm>;
    loadRealmSync: () => Realm;
    reloadRealmSync: () => Realm;

}
export type RealmStackRow = {
    name: string;
    list: Realm.List<RSSnapshot>;
}

type RSSnapshotKeys = typeof COLUMN_NAME_SNAPSHOT_TIMESTAMP;
export type RSSnapshot = Record<RSSnapshotKeys, Date> & Dict<any>;
