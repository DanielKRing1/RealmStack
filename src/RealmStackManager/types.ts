import { RealmStack, RSCreateParams } from "../RealmStack/types";

export type RealmStackManager = {
    getStack: (stackName: string) => RealmStack | undefined;
    createStack: (params: RSCreateParams) => Promise<RealmStack>;
    rmStack: (stackName: string) => void;
    loadStacks: (metaRealmPAth: string, loadableRealmPath: string) => Promise<number>;

    getLoadableStackNames: (metaRealmPath: string, loadableRealmPath: string) => string[];
    getAllLoadedStackNames: () => string[];
    getAllLoadedStacks: () => RealmStack[];
};
