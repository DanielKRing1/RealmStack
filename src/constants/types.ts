export type StackSchemas = {
    snapshotSchema: Realm.ObjectSchema;
    listSchema: Realm.ObjectSchema;
};