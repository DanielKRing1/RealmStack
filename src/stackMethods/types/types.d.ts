import { COLUMN_NAME_SNAPSHOT_TIMESTAMP } from "../constants";

declare global {

    type StackSnapshotRow = Dict<number | string | Date> & {
        [COLUMN_NAME_SNAPSHOT_TIMESTAMP]: Date;
    };
    type StackListRow = {
        name: string;
        list: Realm.List<StackSnapshotRow>;
    };

}
