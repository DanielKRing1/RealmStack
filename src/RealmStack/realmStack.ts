import Realm from 'realm';
import MetaRealm, { SaveSchemaParams } from '@asianpersonn/metarealm';

import { genStackSchemas } from "../constants/schemas";
import { RealmStack, RealmStackRow, RSCreateParams, RSDeleteStackParams, RSLoadStackParams, RSPushParams, RSSnapshot, RSUpdatePropertiesParams } from "./types";
import { COLUMN_NAME_SNAPSHOT_TIMESTAMP, getSnapshotSchemaName, getStackSchemaName, PK_STACK_LIST_ROW } from '../constants/naming';
import { binarySearchClosest } from '../utils/binarySearch';
import { Dict } from '@asianpersonn/dict-utils/dist/types';


export const createRealmStack = async({ metaRealmPath, loadableRealmPath, stackName, snapshotProperties }: RSCreateParams): Promise<RealmStack> => {
    // 1. Save new Stack Snapshot and List schemas
    _saveStackSchemas(metaRealmPath, loadableRealmPath, stackName, snapshotProperties);
    
    // 2. Reload LoadableRealm with new schemas, so propertyNames can be accessed from LoadableSchema
    await MetaRealm.LoadableRealmManager.reloadRealm({ metaRealmPath, loadableRealmPath });

    // 3. Add defauly StackList row
    await _initStackListRow(metaRealmPath, loadableRealmPath, stackName);

    // 4. Setup RealmStack
    const realmGraph: RealmStack = initializeRealmStack({ metaRealmPath, loadableRealmPath, stackName, snapshotProperties });

    return realmGraph;
}

export const loadRealmStack = async ({ metaRealmPath, loadableRealmPath, stackName, shouldReloadRealm=true }: RSLoadStackParams): Promise<RealmStack> => {
    // 1. Get propertyNames from saved Node schema
    // Can use Node or Edge schema
    const snapshotSchemaName: string = getSnapshotSchemaName(stackName);
    const snapshotProperties: Realm.PropertiesTypes = MetaRealm.getProperties(metaRealmPath, snapshotSchemaName);
    
    // 2. Remove unrelated keys
    delete snapshotProperties[COLUMN_NAME_SNAPSHOT_TIMESTAMP];
    
    // 3. Initialize RealmStack
    const realmStack: RealmStack = await initializeRealmStack({ metaRealmPath, loadableRealmPath, stackName, snapshotProperties });

    // 4. Reload realm (optional bcus may be loading multiple RealmStqacks)
    if(shouldReloadRealm) await realmStack.reloadRealm();
    
    return realmStack;
};

const initializeRealmStack = ({ metaRealmPath, loadableRealmPath, stackName }: RSCreateParams): RealmStack => {

    // REALM GRAPH METHODS
    function getStackName(): string { return stackName; };
    function getMetaRealmPath(): string { return metaRealmPath; };
    function getLoadableRealmPath(): string { return loadableRealmPath; };
    function getPropertyNames(): string[] { return Object.keys(getProperties()); };
    function getProperties(): Dict<any> {
        const properties: Dict<any> = MetaRealm.getProperties(metaRealmPath, getSnapshotSchemaName(stackName));
        delete properties[COLUMN_NAME_SNAPSHOT_TIMESTAMP];

        return properties;
    };

    const getAllSnapshots = async (): Promise<Realm.Results<RSSnapshot & Realm.Object>> => {
        const loadedGraphRealm: Realm = await loadRealm();
        return loadedGraphRealm.objects(getSnapshotSchemaName(stackName));
    }

    /**
     * Get the Stack row Realm.Result
     * 
     * @returns 
     */
    const getStackRow = async (): Promise<(RealmStackRow & Realm.Object) | undefined> => {
        const loadedGraphRealm: Realm = await loadRealm();
        const stackRow: (RealmStackRow & Realm.Object) | undefined = loadedGraphRealm.objectForPrimaryKey(getStackSchemaName(stackName), PK_STACK_LIST_ROW);

        return stackRow;
    }

    /**
     * Get the JSON-ified stack as a list
     * 
     * @returns 
     */
    const getListJSON = async (): Promise<RSSnapshot[]> => {
        const stackRow: (RealmStackRow & Realm.Object) | undefined = await getStackRow();

        try {
            return stackRow !== undefined ? stackRow.list.toJSON() : [];
        }
        catch(err) {
            console.log(err);
        }
    }

    const deleteStack = async (): Promise<void> => await _deleteStackSchemas({ metaRealmPath, loadableRealmPath, stackName, reloadRealm });

    const updateStack = async (newSnapshotProperties: Dict<any>): Promise<void> => await _updateStackSchemas({ metaRealmPath, loadableRealmPath, stackName, newSnapshotProperties, reloadRealm });

    // PUSH

    /**
     * Push a list of snapshots onto the RealmStack
     * Returns true if snapshots are successfully added
     * 
     * @param param0 
     * @param snapshots 
     * @returns 
     */
    const push = async ({ timestamp = new Date() }: RSPushParams, ...snapshots: RSSnapshot[]): Promise<boolean> => {
        // 1. Get realm
        const realm: Realm = await MetaRealm.LoadableRealmManager.loadRealm({ metaRealmPath, loadableRealmPath });

        // 2. Get Stack
        const stack: (RealmStackRow & Realm.Object) | undefined = await getStackRow();

        // 3. Append timestamp to snapshot
        const formatedSnapshots: RSSnapshot[] = snapshots.map((snapshot: Dict<any>) => ({
            ...snapshot,
            [COLUMN_NAME_SNAPSHOT_TIMESTAMP]: timestamp,
        }));

        // 4. Add snapshots to front of stack
        realm.write(() => {
            stack.list.unshift(...formatedSnapshots);
        });

        return true;
    }

    // SEARCH
    
    /**
     * Returns the index of the oldest snapshot gte to the search date
     * 
     * @param searchDate 
     * @returns 
     */
    const getClosestDate = async (searchDate: Date): Promise<number> => {
        // 1. Get Stack
        const list: RSSnapshot[] = await getListJSON();

        if(list.length === 0 ) return -1;

        // EDGE CASES
        // [ newest time, ..., oldest time ] (Descending order)

        // 2. No dates gte searchDate
        // Search date is more recent than latest snapshot
        if (searchDate < list[list.length - 1][COLUMN_NAME_SNAPSHOT_TIMESTAMP]) return list.length - 1;

        // 3. All dates gte searchDate
        // Search date is older than oldest snapshot
        if (searchDate > list[0][COLUMN_NAME_SNAPSHOT_TIMESTAMP]) return 0;

        // 4. Binary search
        return _getDateInStackLTE(searchDate, list);
    };

    // INTERNAL UTILITY
    async function loadRealm(): Promise<Realm> { return await MetaRealm.LoadableRealmManager.loadRealm({ metaRealmPath, loadableRealmPath }) };
    async function reloadRealm(): Promise<Realm> { return await MetaRealm.LoadableRealmManager.reloadRealm({ metaRealmPath, loadableRealmPath }) };
    function loadRealmSync(): Realm { return MetaRealm.LoadableRealmManager.loadRealmSync({ metaRealmPath, loadableRealmPath }) };
    function reloadRealmSync(): Realm { return MetaRealm.LoadableRealmManager.reloadRealmSync({ metaRealmPath, loadableRealmPath }) };

    return {
        getStackName,
        getMetaRealmPath,
        getLoadableRealmPath,
        getPropertyNames,
        getProperties,

        getAllSnapshots,
        getStackRow,
        getListJSON,

        deleteStack,
        updateStack,

        push,
        getClosestDate,

        loadRealm,
        reloadRealm,
        loadRealmSync,
        reloadRealmSync,
    }
}

const _saveStackSchemas = (metaRealmPath: string, loadableRealmPath: string, stackName: string, snapshotProperties: Dict<any>): void => {
    const { snapshotSchema, listSchema } = genStackSchemas(stackName, snapshotProperties);

    const saveParams: SaveSchemaParams[] = [snapshotSchema, listSchema].map((schema: Realm.ObjectSchema) => ({ metaRealmPath, loadableRealmPath, schema }));
    // 2. And save new Graph schemas
    for(let saveParam of saveParams) {
        MetaRealm.saveSchema(saveParam);
    }
}

const _initStackListRow = async (metaRealmPath: string, loadableRealmPath: string, stackName: string): Promise<void> => {
    // 1. Create stack row in Stack schema
    const realm: Realm = await MetaRealm.LoadableRealmManager.loadRealm({ metaRealmPath, loadableRealmPath });
    try {
        realm.write(() => {
            realm.create(getStackSchemaName(stackName), {
                name: PK_STACK_LIST_ROW,
                list: [],
            });
        });
    } catch (err) {
        // Error thrown to prevent writing duplicate; Do nothing
        console.log(err);
    }
}

const _deleteStackSchemas = async ({ metaRealmPath, loadableRealmPath, stackName, reloadRealm }: RSDeleteStackParams): Promise<void> => {
    // 1. Get node + edge schema names
    const schemaNames: string[] = [ getSnapshotSchemaName(stackName), getStackSchemaName(stackName) ];

    // 2. Delete Graph schemas
    for(let schemaName of schemaNames) {
        MetaRealm.rmSchema({ metaRealmPath, loadableRealmPath, schemaName });
    }

    // 3. Reload Realm with updated schemas
    await reloadRealm();
}

const _updateStackSchemas = async ({ metaRealmPath, loadableRealmPath, stackName, newSnapshotProperties, reloadRealm }: RSUpdatePropertiesParams): Promise<void> => {
    // 1. Create updated stack schemas (only snapshot schema))
    const { snapshotSchema } = genStackSchemas(stackName, newSnapshotProperties);

    const updatedSchemas: Realm.ObjectSchema[] = [ snapshotSchema ];
    // 2. Save updated RealmStack schema (only snapshot schema)
    for(let schema of updatedSchemas) {
        MetaRealm.updateSchema({ metaRealmPath, loadableRealmPath, schema });
    }

    // 3. Reload Realm with updated schemas
    await reloadRealm();
}


/**
 * Returns the index of the oldest snapshot gte to the search date
 * 
 * @param searchDate 
 * @param list 
 * @returns 
 */
const _getDateInStackLTE = (searchDate: Date, list: RSSnapshot[]): number => {
    if(list.length === 0 ) return -1;

    const comparator = (midValue: RSSnapshot, searchDate: Date): -1 | 0 | 1 => {
        if(searchDate > midValue[COLUMN_NAME_SNAPSHOT_TIMESTAMP]) return -1;
        else if (searchDate < midValue[COLUMN_NAME_SNAPSHOT_TIMESTAMP]) return 1;
        else return 0;
    };

    // 1. Get index at which dates to the left are gte
    const foundIndex: number = binarySearchClosest<RSSnapshot>(searchDate, list, comparator);

    // 2. If exact date is found during binary search
    //      and there are duplicates of this date in the stack,
    //      then might not return index for right-most duplicate
    // Check to the right of list for first instance of (duplicate) date
    for(let oldestDuplicateIndex = foundIndex + 1; oldestDuplicateIndex < list.length; oldestDuplicateIndex++) {
        const snapshot: RSSnapshot = list[oldestDuplicateIndex];
        const timestamp: Date = snapshot[COLUMN_NAME_SNAPSHOT_TIMESTAMP];

        // Not a duplicate, so return last duplicate index (currentIndex - 1)
        if(list[foundIndex][COLUMN_NAME_SNAPSHOT_TIMESTAMP].getTime() != timestamp.getTime()) return oldestDuplicateIndex - 1;
    }

    // Contains duplicates until the last index
    return list.length - 1;
}
