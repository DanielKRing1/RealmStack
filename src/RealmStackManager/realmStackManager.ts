import { Dict } from '@asianpersonn/dict-utils/dist/types';
import MetaRealm from '@asianpersonn/metarealm';
import { getBaseNameFromSchemaName } from '../constants/naming';

import { createRealmStack, loadRealmStack } from "../RealmStack/realmStack";
import { RealmStack, RSCreateParams } from "../RealmStack/types";
import { gen_NO_STACK_ERROR } from './errors';
import { RealmStackManager } from './types';

const createStackManager = (): RealmStackManager => {
    const realmStackMap: Dict<RealmStack> = {};
    
    const getStack = (stackName: string): RealmStack | never => {
        if(!hasRealmStack(stackName)) throw gen_NO_STACK_ERROR(stackName);
        
        return realmStackMap[stackName];
    }
    
    const createStack = async ({ metaRealmPath, loadableRealmPath, stackName, snapshotProperties }: RSCreateParams) => {
        // 1. Create new RealmStack if not exists
        if(!hasRealmStack(stackName)) {
            const realmStack: RealmStack = await createRealmStack({
                metaRealmPath,
                loadableRealmPath,
                stackName,
                snapshotProperties,
            });
            realmStackMap[stackName] = realmStack;
        }

        return realmStackMap[stackName];
    }
    const rmStack = async (stackName: string) => {
        // 1. Try to delete RealmStack schemas
        if(hasRealmStack(stackName)) {
            const realmStack: RealmStack = realmStackMap[stackName];
            await realmStack.deleteStack();
        }

        // 2. Remove RealmGraph from map if present
        delete realmStackMap[stackName];
    }

    const hasRealmStack = (stackName: string): boolean => !!realmStackMap[stackName];

    const loadStacks = async (metaRealmPath: string, loadableRealmPath: string) => {
      // 1. Get all Loadable graph names
      const stackNames: string[] = getLoadableStackNames(metaRealmPath, loadableRealmPath);

      for(let stackName of stackNames) {
          // 2. Setup RealmGraph, It will read its properties from its LoadableRealm schemas
          const realmStack: RealmStack = await loadRealmStack({
              metaRealmPath,
              loadableRealmPath,
              stackName,
              shouldReloadRealm: false,
          });

          // 3. Add to realmStackMap
          realmStackMap[stackName] = realmStack;
      }

      // 4. Reload realms after loading all RealmStacks
      await MetaRealm.LoadableRealmManager.reloadRealm({ metaRealmPath, loadableRealmPath });

      return stackNames.length;
    }

    const getLoadableStackNames = (metaRealmPath: string, loadableRealmPath: string): string[] => {
        const stackNames: Set<string> = new Set<string>();

        // 1. Get all schema names
        const allSchemaNames: string[] = MetaRealm.getSchemaNames(metaRealmPath, loadableRealmPath);

        // 2. Remove schema name suffix and add to set
        // (A single Graph creates more than 1 schema; when stripped of their suffixes, they will have identical names)
        allSchemaNames.forEach((schemaName) => {
            // 2.1. Remove suffix
            const stackName: string = getBaseNameFromSchemaName(schemaName);

            // 2.2. Add to set
            stackNames.add(stackName);
        });

        return Array.from(stackNames);
    }

    const getAllLoadedStackNames = () => Object.keys(realmStackMap);
    const getAllLoadedStacks = () => Object.values(realmStackMap);

    return {
        getStack,
        createStack,
        rmStack,
        loadStacks,

        getLoadableStackNames,
        getAllLoadedStackNames,
        getAllLoadedStacks,
    }
}

export default createStackManager();
