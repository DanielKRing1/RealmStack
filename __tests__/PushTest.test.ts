import fs from 'fs';

import MetaRealm from '@asianpersonn/metarealm';
import { createRealmStack } from '../src/RealmStack/realmStack';
import { RealmStack } from '../src/RealmStack/types';
import realmStackManager from '../src/RealmStackManager/realmStackManager';
import { Dict } from '@asianpersonn/dict-utils/dist/types';

const TEST_NAME: string = 'PushTest';
const TEST_DIRECTORY: string = `__tests__/${TEST_NAME}`;
const META_REALM_PATH1: string = `${TEST_DIRECTORY}/MetaRealm1.path`;
const LOADABLE_REALM_PATH1: string = `LoadableRealm1.path`;

const STACK_NAME1: string = 'TestStack1';
const STACK_NAME2: string = 'TestStack2';

const PROPERTIES1: Dict<any> = {
    'prop1': 'int',
    'prop2': 'string',
};
const PROPERTIES2: Dict<any> = {
    'prop3': 'int',
    'prop4': 'date',
};

describe('RealmStackManager', () => {
    beforeAll(() => {
        if (fs.existsSync(TEST_DIRECTORY)) fs.rmSync(TEST_DIRECTORY, { recursive: true });

        fs.mkdirSync(TEST_DIRECTORY);
    });

    it('Should load all stacks in same realm', async () => {
        await createRealmStack({
            metaRealmPath: META_REALM_PATH1,
            loadableRealmPath: LOADABLE_REALM_PATH1,
            stackName: STACK_NAME1,
            snapshotProperties: PROPERTIES1,
        });
        await createRealmStack({
            metaRealmPath: META_REALM_PATH1,
            loadableRealmPath: LOADABLE_REALM_PATH1,
            stackName: STACK_NAME2,
            snapshotProperties: PROPERTIES2,
        });

        await realmStackManager.loadStacks(META_REALM_PATH1, LOADABLE_REALM_PATH1);
    });

    it('Should throw an error for a non-existant RealmStack name', async () => {
        const NONEXISTANT_STACK_NAME: string = 'DUMMY_STACK_NAME';
        expect(() => realmStackManager.getStack(NONEXISTANT_STACK_NAME)).toThrowError();
    });

    it('Should push onto Stack1', async () => {
        const stack: RealmStack = realmStackManager.getStack(STACK_NAME1);

        await stack.push({ timestamp: new Date(Date.parse('9/1/2022')) }, { 'prop1': 0, 'prop2': 'Hello' });
        await stack.push({ timestamp: new Date(Date.parse('9/2/2022')) }, { 'prop1': 0, 'prop2': 'World' });

        await stack.push({ timestamp: new Date(Date.parse('9/3/2022')) }, { 'prop1': 1, 'prop2': 'Hello2' });
        await stack.push({ timestamp: new Date(Date.parse('9/4/2022')) }, { 'prop1': 1, 'prop2': 'World2' });

        await stack.push({ timestamp: new Date(Date.parse('9/5/2022')) }, { 'prop1': 2, 'prop2': 'Hello3' });
        await stack.push({ timestamp: new Date(Date.parse('9/6/2022')) }, { 'prop1': 3, 'prop2': 'World3' });

        await stack.push({ timestamp: new Date(Date.parse('9/7/2022')) }, { 'prop1': 4, 'prop2': 'Hello4' });
        await stack.push({ timestamp: new Date(Date.parse('9/8/2022')) }, { 'prop1': 5, 'prop2': 'World4' });
    });

    it('Should push onto Stack2', async () => {
        const stack: RealmStack = realmStackManager.getStack(STACK_NAME2);

        await stack.push({ timestamp: new Date(Date.parse('9/1/2022')) }, { 'prop3': 0, 'prop4': new Date(Date.parse('9/10/2022')) });
        await stack.push({ timestamp: new Date(Date.parse('9/2/2022')) }, { 'prop3': 0, 'prop4': new Date(Date.parse('9/10/2022')) });

        await stack.push({ timestamp: new Date(Date.parse('9/3/2022')) }, { 'prop3': 1, 'prop4': new Date(Date.parse('9/10/2022')) });
        await stack.push({ timestamp: new Date(Date.parse('9/4/2022')) }, { 'prop3': 1, 'prop4': new Date(Date.parse('9/10/2022')) });

        await stack.push({ timestamp: new Date(Date.parse('9/5/2022')) }, { 'prop3': 2, 'prop4': new Date(Date.parse('9/10/2022')) });
        await stack.push({ timestamp: new Date(Date.parse('9/6/2022')) }, { 'prop3': 3, 'prop4': new Date(Date.parse('9/10/2022')) });

        await stack.push({ timestamp: new Date(Date.parse('9/7/2022')) }, { 'prop3': 4, 'prop4': new Date(Date.parse('9/10/2022')) });
        await stack.push({ timestamp: new Date(Date.parse('9/8/2022')) }, { 'prop3': 5, 'prop4': new Date(Date.parse('9/10/2022')) });
    });

    it('Should get lists from Stacks 1-2', async () => {
        // Stack1
        const stack1: RealmStack = realmStackManager.getStack(STACK_NAME1);

        expect((await stack1.getAllSnapshots()).toJSON()).toMatchSnapshot();
        expect((await stack1.getListJSON())).toMatchSnapshot();

        // Stack2
        const stack2: RealmStack = realmStackManager.getStack(STACK_NAME2);

        expect((await stack2.getAllSnapshots()).toJSON()).toMatchSnapshot();
        expect((await stack2.getListJSON())).toMatchSnapshot();
    });

    afterAll(async () => {
        await MetaRealm.MetaRealmManager.closeAll();
        await MetaRealm.LoadableRealmManager.closeAll();

        if (fs.existsSync(TEST_DIRECTORY)) fs.rmSync(TEST_DIRECTORY, { recursive: true });
    });
});
