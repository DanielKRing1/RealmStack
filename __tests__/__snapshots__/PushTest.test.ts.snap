// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RealmStackManager Should get lists from Stacks 1-2 1`] = `
Array [
  Object {
    "prop1": 0,
    "prop2": "Hello",
    "timestamp": 2022-09-01T07:00:00.000Z,
  },
  Object {
    "prop1": 0,
    "prop2": "World",
    "timestamp": 2022-09-02T07:00:00.000Z,
  },
  Object {
    "prop1": 1,
    "prop2": "Hello2",
    "timestamp": 2022-09-03T07:00:00.000Z,
  },
  Object {
    "prop1": 1,
    "prop2": "World2",
    "timestamp": 2022-09-04T07:00:00.000Z,
  },
  Object {
    "prop1": 2,
    "prop2": "Hello3",
    "timestamp": 2022-09-05T07:00:00.000Z,
  },
  Object {
    "prop1": 3,
    "prop2": "World3",
    "timestamp": 2022-09-06T07:00:00.000Z,
  },
  Object {
    "prop1": 4,
    "prop2": "Hello4",
    "timestamp": 2022-09-07T07:00:00.000Z,
  },
  Object {
    "prop1": 5,
    "prop2": "World4",
    "timestamp": 2022-09-08T07:00:00.000Z,
  },
]
`;

exports[`RealmStackManager Should get lists from Stacks 1-2 2`] = `
Array [
  Object {
    "prop1": 5,
    "prop2": "World4",
    "timestamp": 2022-09-08T07:00:00.000Z,
  },
  Object {
    "prop1": 4,
    "prop2": "Hello4",
    "timestamp": 2022-09-07T07:00:00.000Z,
  },
  Object {
    "prop1": 3,
    "prop2": "World3",
    "timestamp": 2022-09-06T07:00:00.000Z,
  },
  Object {
    "prop1": 2,
    "prop2": "Hello3",
    "timestamp": 2022-09-05T07:00:00.000Z,
  },
  Object {
    "prop1": 1,
    "prop2": "World2",
    "timestamp": 2022-09-04T07:00:00.000Z,
  },
  Object {
    "prop1": 1,
    "prop2": "Hello2",
    "timestamp": 2022-09-03T07:00:00.000Z,
  },
  Object {
    "prop1": 0,
    "prop2": "World",
    "timestamp": 2022-09-02T07:00:00.000Z,
  },
  Object {
    "prop1": 0,
    "prop2": "Hello",
    "timestamp": 2022-09-01T07:00:00.000Z,
  },
]
`;

exports[`RealmStackManager Should get lists from Stacks 1-2 3`] = `
Array [
  Object {
    "prop3": 0,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-01T07:00:00.000Z,
  },
  Object {
    "prop3": 0,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-02T07:00:00.000Z,
  },
  Object {
    "prop3": 1,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-03T07:00:00.000Z,
  },
  Object {
    "prop3": 1,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-04T07:00:00.000Z,
  },
  Object {
    "prop3": 2,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-05T07:00:00.000Z,
  },
  Object {
    "prop3": 3,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-06T07:00:00.000Z,
  },
  Object {
    "prop3": 4,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-07T07:00:00.000Z,
  },
  Object {
    "prop3": 5,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-08T07:00:00.000Z,
  },
]
`;

exports[`RealmStackManager Should get lists from Stacks 1-2 4`] = `
Array [
  Object {
    "prop3": 5,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-08T07:00:00.000Z,
  },
  Object {
    "prop3": 4,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-07T07:00:00.000Z,
  },
  Object {
    "prop3": 3,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-06T07:00:00.000Z,
  },
  Object {
    "prop3": 2,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-05T07:00:00.000Z,
  },
  Object {
    "prop3": 1,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-04T07:00:00.000Z,
  },
  Object {
    "prop3": 1,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-03T07:00:00.000Z,
  },
  Object {
    "prop3": 0,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-02T07:00:00.000Z,
  },
  Object {
    "prop3": 0,
    "prop4": 2022-09-10T07:00:00.000Z,
    "timestamp": 2022-09-01T07:00:00.000Z,
  },
]
`;
